### PR TITLE
Components: Add SubMasterbarNav component

### PR DESCRIPTION
**Pull request by robgietema**
Add a component to display a header navigation bar.

This navigation bar will be used on the themes page to filter between different categories. The navigation bar will be responsive and will either show all buttons (desktop), fold the buttons which don't fit (tablet) or display a dropdown.

header-nav

### CODE REVIEW

**Comment by ice9js** [comment]
I added the component to theme-showcase page for not logged users.
The icons are pulled from the ```subject:``` filter based on the subjects meta-data I defined for the subjects that should be displayed in the header. All subjects missing from the list get filtered out in the header.

I'm currently working on styling the header and making it responsive, here's a preview:

header-nav

**Comment by ice9js** [comment]
The mobile dropdown is now done and looks like this:

headernav-mobile

The currently selected item is not displayed on the list as it's already on the top. The dropdown is hidden as soon as a new subject is selected or a user starts scrolling.
I also made sure the list or any overflow icons are always hidden when passing a breakpoint.

**Comment by ice9js** [comment]
In order to make the header fully functional, I'll need to expose a function that can produce the right filter strings for the url [here](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/themes/theme-filters.js). (For example, portfolio needs to be prefixed with ```subject:```).  
I'll open a separate pull request for that.

The reason I decided to stick with links is because it seems the most universal approach, should any other section need such a component.

**Comment by ockham** [comment]
> In order to make the header fully functional, I need to generate urls for each filter taking into account whether the terms are ambiguous or not. For example, portfolio requires the ```subject:``` prefix but the rest doesn't.  
> I'm thinking about adding and exposing such a function [here](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/themes/theme-filters.js).

Yeah, that ambiguity kind of sucks. I thought we had a function like that already somewhere, since we also needed it for our (new, "magic") search bar?

> The reason I decided to stick with links is because it seems the most universal approach, should any other section need such a component.

Links are perfectly fine, we also want them for SEO purposes.

**Comment by folletto** [comment]
I gave it a run on Chrome.

Notes:

1. Some colors are a bit off (using probably the blue instread of a shade of gray) but I'd like to review that, so in this regard let's keep it this way for this PR, and iterate in a follow-up PR. I need to review it.
2. "New" should be highlighted on first load if the URL is the base url.
3. The dropdown shouldn't remove the selected item from the list when selected, as the disappearance is equivalent of moving items. Let's just replicate it, keeping it in the list too.

So design wise, with these two changes (2 and 3) we should be good for this PR 👍

File: client/components/header-nav/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Header Nav
```
Comment (mtias) [inline client/components/header-nav/README.md]: I feel like this should be named differently. To me it means "Masterbar Secondary Navigation", but that's a bit long :)

Comment (folletto) [inline client/components/header-nav/README.md]: Submasterheader! :D 

j/k.

Something like: `SubHeader`?

File: client/components/header-nav/index.jsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+ */
+import React, { Component, PropTypes } from 'react';
+import { findDOMNode } from 'react-dom';
+import classNames from 'classnames';
+
+import Gridicon from 'gridicons';
+import Item from 'components/header-nav/item';
+
+export default class HeaderNav extends Component {
+	static propTypes = {
```
Comment (seear) [inline client/components/header-nav/index.jsx]: missing `href` prop

File: client/components/header-nav/index.jsx [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+	static defaultProps = {
+		options: []
+	}
+
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			selected: props.href,
```
Comment (seear) [inline client/components/header-nav/index.jsx]: Does this need to be state?

Comment (ice9js) [inline client/components/header-nav/index.jsx]: You're right, now that the current URI is passed as a property there is no need for ```state.selected```.

File: client/components/header-nav/index.jsx [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	hasChangedView( oldWidth, newWidth, options ) {
+		const min = Math.min( oldWidth, newWidth );
+		const max = Math.max( oldWidth, newWidth );
+
+		return ( this.isMobile( min ) && ! this.isMobile( max ) ) ||
+			( this.isFoldable( min, options ) && ! this.isFoldable( max, options ) );
+	}
+
+	isFoldable( width, options ) {
+		return 660 < width && width < options.length * 110 + 50;
```
Comment (ehg) [inline client/components/header-nav/index.jsx]: These numbers seem a little magic.

File: client/components/header-nav/index.jsx [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	static defaultProps = {
+		options: []
+	}
+
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			collapsed: true,
+			width: 99999
```
Comment (ehg) [inline client/components/header-nav/index.jsx]: Why 99999? Also, the state could be set via an ESNext property initialiser.

File: client/components/header-nav/test/index.jsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import { expect } from 'chai';
+import React from 'react';
+import { shallow } from 'enzyme';
+
+/**
+ * Internal dependencies
+ */
+import useFakeDom from 'test/helpers/use-fake-dom';
+
+describe( 'HeaderNav', function() {
```
Comment (ehg) [inline client/components/header-nav/test/index.jsx]: These tests seem a little anaemic.

**Comment by ehg** [comment]
Once you're ready. I think it'd be better to split up the the PR into the generic component, and the implementation with themes-logged-out. What do you think?

File: client/components/header-nav/index.jsx [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+		return ( this.isMobile( min ) && ! this.isMobile( max ) ) ||
+			( this.isFoldable( min, options ) && ! this.isFoldable( max, options ) );
+	}
+
+	isFoldable( width, options ) {
+		return 660 < width && width < options.length * 110 + 50;
+	}
+
+	isMobile( width ) {
```
Comment (mtias) [inline client/components/header-nav/index.jsx]: This function is not really checking for mobile. Could we name it differently?

File: client/components/header-nav/index.jsx [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+		return (
+			<div className={ className }>
+				{ this.isMobile( this.state.width ) &&
+					<div className="header-nav__select" onClick={ this.toggleList.bind( this ) }>
+						<Item
+							isSelected={ true }
+							label={ this.getSelected().label }
+							icon={ this.getSelected().icon }
+						/>
+						<Gridicon icon="chevron-down" className="header-nav__select__icon" />
```
Comment (mtias) [inline client/components/header-nav/index.jsx]: This class should (hopefully) be giving a warning.

File: client/components/header-nav/README.md (hunk 1 above)
Comment (ice9js) [inline client/components/header-nav/README.md]: How about ```SectionNavbar``` ? Does that make sense ?

Comment (ehg) [inline client/components/header-nav/README.md]: We already have a `SectionNav` :/ How about `SubMasterBarNav`? ;) I'm sure we've had this discussion on Slack a year ago!

Comment (folletto) [inline client/components/header-nav/README.md]: Just `SubMasterbar`?

Comment (ice9js) [inline client/components/header-nav/README.md]: Another suggestion, ```FoldingNavbar```. What do you guys think ?

Comment (folletto) [inline client/components/header-nav/README.md]: Why "folding"?

Comment (ice9js) [inline client/components/header-nav/README.md]: Because of how it folds when the viewport gets smaller. Just an idea.

Comment (folletto) [inline client/components/header-nav/README.md]: Ah ok. That's just a way to handle responsiveness, I wouldn't make it a characteristic trait. I think its defining properties are being a navigation item, being attached to the masterbar. :)

Comment (ice9js) [inline client/components/header-nav/README.md]: Alright, I'm going with ```SubMasterbarNav```.  
Using just ```SubMasterbar```might prove problematic in the future in case we'd want to put  something else up there.

**Comment by ice9js** [comment]
As suggested by @ehg , I split the PR into two parts.  

Part one (this PR) adds the ```SubMasterbarNav``` component.  
Part two (#12585) adds the component to ```theme-showcase```.

File: client/components/sub-masterbar-nav/index.jsx [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+	hasChangedView( oldWidth, newWidth, options ) {
+		const min = Math.min( oldWidth, newWidth );
+		const max = Math.max( oldWidth, newWidth );
+
+		return ( this.isDroppable( min ) && ! this.isDroppable( max ) ) ||
+			( this.isFoldable( min, options ) && ! this.isFoldable( max, options ) );
+	}
+
+	isFoldable( width, options ) {
+		return VIEW_BREAKPOINT < width && width < options.length * ITEM_WIDTH + VIEW_PADDING;
```
Comment (ehg) [inline client/components/sub-masterbar-nav/index.jsx]: Could https://github.com/Automattic/wp-calypso/blob/master/client/lib/viewport/index.js#L30 be useful here?

Comment (ice9js) [inline client/components/sub-masterbar-nav/index.jsx]: I'm not actually sure it would as it pulls the width from the window instead of taking a parameter.  
This would cause problems for ```hasChangedView``` which collapses the dropdown/list whenever the breakpoint is crossed.

If we don't care about that, I can remove that function and use ```isWithinBreakpoint``` where appropriate.

File: client/components/sub-masterbar-nav/index.jsx [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	state = {
+		collapsed: true
+	};
+
+	componentDidMount() {
+		this.onResize();
+
+		window.addEventListener( 'resize', this.onResize.bind( this ) );
```
Comment (budzanowski) [inline client/components/sub-masterbar-nav/index.jsx]: I always get confused by `=>` functions. What if you would declare `onResize` with arrow function? Would you still need bind here?

File: client/components/sub-masterbar-nav/index.jsx [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	hasChangedView( oldWidth, newWidth, options ) {
+		const min = Math.min( oldWidth, newWidth );
+		const max = Math.max( oldWidth, newWidth );
+
+		return ( this.isDroppable( min ) && ! this.isDroppable( max ) ) ||
+			( this.isFoldable( min, options ) && ! this.isFoldable( max, options ) );
+	}
+
```
Comment (budzanowski) [inline client/components/sub-masterbar-nav/index.jsx]: Could you add 1 or 2 lines of comment about what `isFoldable` and `isDropable` are doing?

File: client/components/sub-masterbar-nav/index.jsx (hunk 10 above)
Comment (ice9js) [inline client/components/sub-masterbar-nav/index.jsx]: That makes sense. Should I do this just for the functions that need it or all of them for better readability ?

Also, how should I solve the issue with ```onClick={ this.selectItem.bind( this, item ) }``` ?  
I have a feeling I should either be passing the whole object to the ```Item``` component and have it pass it back as an argument, or I should be passing indexes instead of options into ```selectItem```.

File: client/components/sub-masterbar-nav/README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
+Sub Masterbar Nav
+=================
+
+This component displays a navigation header under the masterbar.
+
+### How to use:
+
+The component should be inserted into a ```Main``` component.
```
Comment (ockham) [inline client/components/sub-masterbar-nav/README.md]: Not too relevant, but do we really need triple backticks everywhere? Aren't single ones enough? :smile:

File: client/components/sub-masterbar-nav/index.jsx [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+				uri: PropTypes.string.isRequired,
+				icon: PropTypes.string
+			}
+		),
+		options: PropTypes.arrayOf(
+			PropTypes.shape( {
+				label: PropTypes.string.isRequired,
+				uri: PropTypes.string.isRequired,
+				icon: PropTypes.string
+			} )
```
Comment (ockham) [inline client/components/sub-masterbar-nav/index.jsx]: Might make sense to define this as an `OptionShape` `const` and re-use both here and for `fallback`.

File: client/components/sub-masterbar-nav/README.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+Options should contain a ```label``` and an ```uri```. If the latter matches the ```uri``` property of the component, it will be marked as selected.  
+In case no option is selected, the ```fallback``` value will be shown in the dropdown.
+
+```icon``` should be a valid gridicon name. The icon will be displayed next to/on top of the label.  
+If no ```icon``` is provided, ```star``` will be used by default.
+
+```js
+const SubMasterbarNav = require( 'components/sub-masterbar-nav' );
+
+const options = {
```
Comment (ockham) [inline client/components/sub-masterbar-nav/README.md]: Nitpick -- should be an array (not an object) to be consistent with the implementation, and https://github.com/Automattic/wp-calypso/pull/11903/files#diff-6e3d613e7ebf414e5f0aadde79da1794R32.

File: client/components/sub-masterbar-nav/index.jsx (hunk 10 above)
Comment (ockham) [inline client/components/sub-masterbar-nav/index.jsx]: > Should I do this just for the functions that need it or all of them for better readability ?

We normally do it just for the ones that need it.

File: client/components/sub-masterbar-nav/index.jsx [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+	isSelected( option ) {
+		return option.uri === this.props.uri;
+	}
+
+	getSelected() {
+		return this.props.options.reduce(
+			( selected, current ) => this.isSelected( current ) ? current : selected,
+			this.props.fallback
+		);
```
Comment (ockham) [inline client/components/sub-masterbar-nav/index.jsx]: Using `reduce` to find something is rather odd. Can't we just use lodash's `find`?

File: client/components/sub-masterbar-nav/index.jsx [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+		if ( ! option ) {
+			return;
+		}
+
+		this.setState( ( state ) => ( {
+			collapsed: this.isDroppable( state.width ) || state.collapsed
+		} ) );
+	}
+
+	toggleList() {
```
Comment (ockham) [inline client/components/sub-masterbar-nav/index.jsx]: Better use [this style](https://babeljs.io/blog/2015/06/07/react-on-es6-plus) (in case that's not obvious from [my other comment](https://github.com/Automattic/wp-calypso/pull/11903/files#r109200435) 😄 ):

```js
class PostInfo extends React.Component {
  handleOptionsButtonClick = (e) => {
    this.setState({showOptionsModal: true});
  }
}
```

**Comment by ice9js** [comment]
I split the component into two additional subcomponents (Dropdown and Navbar), and got rid of other code smells.

I still have to figure something out about server-side rendering, as it won't work the way it is now.

File: client/components/sub-masterbar-nav/dropdown.jsx [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Internal dependencies
+ */
+import Gridicon from 'gridicons';
+import Item from './item';
+
+export default class Dropdown extends Component {
+
+	static propTypes = {
+		children: PropTypes.func.isRequired,
```
Comment (budzanowski) [inline client/components/sub-masterbar-nav/dropdown.jsx]: I have hard time following this. I will explain my line of thoughts on how I would see it so you can comment.
I would rather not pass callback that generates children but the actual data that are used to generate children. Then I would implement function that generates children ( Item ) in both components even if it would mean duplication - but in that case not that big. Why? 

* It is hard to figure out ( not that bad here but it is a general remark ) how the passed props are used.
You pass children callback to this component and it is used here. At the same time you pass `onSelect` into the callback, and to understand how it will be used you need to back to the calling component that is the origin of `children` callback.

* This is I think more important: we loose all benefits of smart optimizations of React lifecycle methods. If in calling component set of items will change then `shouldComponentUpdate` here will only see the method in props that will not change. Right now it is not the case here but we don't know how this component will end up.

* Last thing I would not use children here. I would rather think in terms of composition.

File: client/components/sub-masterbar-nav/index.jsx [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+	render() {
+		return (
+			<div className="sub-masterbar-nav">
+				<Dropdown selected={ this.getSelected() }>
+					{ ( { onSelect } ) => (
+						this.renderItems( onSelect )
+					) }
+				</Dropdown>
+				<Navbar>
+					{ this.renderItems( noop ) }
```
Comment (budzanowski) [inline client/components/sub-masterbar-nav/index.jsx]: The previous comment about children callback makes even more sense here. Why have general renderItems if in just second usage you need to pass `noop` when there should be just nothing in more specific render inside `Navbar`

**Comment by budzanowski** [comment]
In general code looks clean. This is getting close to done.

**Comment by budzanowski** [comment]
We have discussed that after the split ( split itself made sense ofc )  it is a bit hard to follow without seeing how it actually works and integrates with the rest of Showcase. Since code looks nice it will be probably useful to push the component to master and focus on it again from the perspective of #12585 - the PR that connects is to the rest of the code. 
@folletto @ehg @seear @ockham any objections? :smiley:

**Comment by folletto** [comment]
I'm not sure I undestand the plan, can you outline it as step-by-step?

**Comment by budzanowski** [comment]
Not really a plan. This PR is just a component that is not connected to showcase, so pushing it to master is a safe operation. Then we will focus on #12585 PR that connects this component to the showcase. We will be able to see how it actually works. We will be able to test it, observe the console and how it performs. We could even consider putting it behind the flag in development environment like we did with M5 to gather additional feedback. While reviewing #12585 we will not only comment the code that connects the component but also the component itself.

**Comment by folletto** [comment]
Seems a good plan to me, but I'll defer to the others being more aware of the architecture to make that call. 👍
